### ISSUE TITLE
[YSQL] Table is in inconsistent state when adding column that is DEFAULT NOT NULL

### ISSUE THREAD

**Issue by lukaseder** [yugabyte/yugabyte-db#9970]
Jira Link: [DB-2570](https://yugabyte.atlassian.net/browse/DB-2570)
I understand that adding columns with a new default doesn't work yet:

```sql
create table t (id int, a varchar(10), primary key (id));
insert into t (id, a) values (1, 'A');
alter table t add b int default 1;
select * from t;
```

Producing:

```
|id |a  |b  |
|---|---|---|
|1  |A  |   |
```

Relevant issues are:

- https://github.com/yugabyte/yugabyte-db/issues/4415
- https://github.com/yugabyte/yugabyte-db/issues/4192

But when adding a `not null` constraint to the column, the statement should be rejected, alas it works:

```sql
create table t (id int, a varchar(10), primary key (id));
insert into t (id, a) values (1, 'A');
alter table t add b int default 1 not null;
select column_name, is_nullable from information_schema.columns where table_name = 't';
```

Producing:

```
|column_name|is_nullable|
|-----------|-----------|
|a          |YES        |
|b          |NO         |
|id         |NO         |
```

I think this is a bug, independent of the above missing features. The following statement sequence where the `DEFAULT` clause and the `NOT NULL` constraint are separated doesn't work:

```sql
create table t (id int, a varchar(10), primary key (id));
insert into t (id, a) values (1, 'A');
alter table t add b int default 1;
alter table t alter b set not null;
```

This produces an error:

> SQL Error [23502]: ERROR: column "b" contains null values

I think that this statement should produce the same error:

```sql
alter table t add b int default 1 not null;
```

**Comment by tedyu** [comment]
I looked at related code for the 'default 1 not null' case.
In AddRelationNewConstraints():
```
        /* If the DEFAULT is volatile we cannot use a missing value */
        if (colDef->missingMode && contain_volatile_functions((Node *) expr))
            colDef->missingMode = false;
```
In the caller, ATExecAddColumn():
```
		rawEnt->missingMode = true;
...
		AddRelationNewConstraints(rel, list_make1(rawEnt), NIL,
...
		if (!rawEnt->missingMode)
			tab->rewrite |= AT_REWRITE_DEFAULT_VAL;
```
Since the default is non-volatile, missingMode field carries value of true.
Later in ATExecAddColumn(), there is check for identity column, DomainHasConstraints and Oid column. None of them is satisfied, leading to AT_REWRITE_DEFAULT_VAL not being added to rewrite.
The result is that, phase 3 of 'alter table', the rewrite, is not performed. Hence there is no error.

This logic is the same in PG (apart from Oid check is gone since Oid is not special column in upstream PG).

**Comment by tedyu** [comment]
If I temporarily drop the contain_volatile_functions() check, I see another error:
```
yugabyte=# alter table t add b int default 1 not null;
ERROR:  Operation not allowed in YugaByte mode heap_insert
```

**Comment by tedyu** [comment]
I was looking at the PG commit:
```
commit 16828d5c0273b4fe5f10f42588005f16b415b2d8
Author: Andrew Dunstan <andrew@dunslane.net>
Date:   Wed Mar 28 10:43:52 2018 +1030

    Fast ALTER TABLE ADD COLUMN with a non-NULL default
```
I added some debug logs in getmissingattr() and slot_getmissingattrs().
None of which shows up when performing 
```
select * from t;
```
Still investigating ...

**Comment by tedyu** [comment]
Query on table created with column having default value returns the default value.

We should consider letting 'alter table x add y default z not null' error out, as one option.
The other option seems to involve storing correct default value for the new column.

**Comment by Gherciu** [comment]
Yep, this behaviour is not fixed, and that is kinda a big problem being honest
Because locally we develop using pg in doker and that works so we expect yugabyte to behave the same

**Comment by fizaaluthra** [comment]
Fixed by commit 0aca19f161ded4a5089d87a5519379794b29cd94
